WalletWasabi/Wallets/TransactionFeeProvider.cs

```diff
@@ -0,0 +1,136 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Logging;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class TransactionFeeProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumRequestsInParallel = 3;
+
+	public TransactionFeeProvider(WasabiHttpClientFactory httpClientFactory)
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler<EventArgs>? RequestedFeeArrived;
+
+	public ConcurrentDictionary<uint256, Money> FeeCache { get; } = new();
+	public ConcurrentQueue<uint256> Queue { get; } = new();
+	private SemaphoreSlim Semaphore { get; } = new(initialCount: 0, maxCount: MaximumRequestsInParallel);
+	private IHttpClient HttpClient { get; }
+
+	private async Task FetchTransactionFeeAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		const int MaxAttempts = 3;
+
+		for (int i = 0; i < MaxAttempts; i++)
+		{
+			try
+			{
+				using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(20));
+				using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+				var response = await HttpClient.SendAsync(
+					HttpMethod.Get,
+					$"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/get-transaction-fee?transactionId={txid}",
+					null,
+					linkedCts.Token).ConfigureAwait(false);
+
+				if (response.StatusCode != HttpStatusCode.OK)
+				{
+					await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				Money fee = await response.Content.ReadAsJsonAsync<Money>().ConfigureAwait(false);
+
+				if (!FeeCache.TryAdd(txid, fee))
+				{
+					throw new InvalidOperationException($"Failed to cache {txid} with fee: {fee}");
+				}
+
+				RequestedFeeArrived?.Invoke(this, EventArgs.Empty);
```

Comment:
RequestedFeeArrived implies that the event will give you some results - even if you will not use it currently give something back. Like the txId and the fee.

Reply:
Why? I can't use it on the UI side.